### PR TITLE
feat(tests): extend BAL create collision test cases

### LINKED ISSUE #2914: Missing BAL test cases of contract creation collision edge cases
When I rewrite the BAL tracker in nimbus EL, I discover that tests-bal@v7.2.0 missing test for this cases:

Condition: a contract creation address collide with existing account, and then the account is deleted or modified.

There are 46 test cases covering balance modification, the balance change appear in the BAL.

But there are zero test cases for the following changes:
1. nonce change
2. code hash change
3. storage change

I have not analyze if the protocol will allow the changes I mentioned.

### PR DESCRIPTION
**Pull request by fselmo**
## 🗒️ Description

- Extend existing collision create2 test to `with_all_create_opcodes`
- Extend create collision tests to check nonce, code, and storage changes, closes #2914

## 🔗 Related Issues or PRs

closes #2914

## ✅ Checklist

- [x] All: Ran fast static checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    just static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-specs/blob/HEAD/docs/) directory.
- [x] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [x] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-specs/blob/HEAD/docs/writing_tests/post_mortems.md) to add an entry the list.
- [x] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-specs/blob/HEAD/tests/static) have been assigned `@ported_from` marker.

#### Cute Animal Picture

Screenshot 2026-05-26 at 08 58 09

### CODE REVIEW

File: tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py [hunk 1]
```diff
@@ -2830,122 +2830,175 @@ def test_bal_create_selfdestruct_to_self_with_call(
...
+            nonce_changes=[BalNonceChange(block_access_index=2, post_nonce=1)],
+        )
+        post[bob] = Account(nonce=1)
+
+    if modification == "none":
+        account_expectations[collision_address] = BalAccountExpectation.empty()
+        post[collision_address] = Account(
+            code=x_code, nonce=1, balance=0, storage={}
+        )
+    elif modification == "nonce_via_inner_create":
```
Comment (raxhvl) [inline tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py]: issue: Dead branches

modification assumes` "collision_only" | "then_nonce_change" | "then_storage_change"`

So comparison to `none`, `nonce_via_inner_create` creates dead branches

Comment (fselmo) [reply tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py]: Thank you! Last minute rename. Will update and do another review before pushing 🙏🏼

Comment (fselmo) [reply tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py]: This is precisely why I hate catch-all conditionals in tests but the quick rename allowed a slip ☹️. Should be fixed now.

File: tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py [hunk 2]
```diff
@@ -2830,122 +2830,175 @@ def test_bal_create_selfdestruct_to_self_with_call(
...
 
 
+@pytest.mark.with_all_create_opcodes
+@pytest.mark.parametrize(
+    "modification",
+    ["collision_only", "then_nonce_change", "then_storage_change"],
+)
 @pytest.mark.pre_alloc_mutable()
-def test_bal_create2_collision(
+def test_bal_create_collision(
```
Comment (raxhvl) [inline tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py:2839]: optional suggestion: Use parameter for branching.

It was kind of tricky to follow the branching for this test. Could we instead use parameter for branching logic. This way we need not handle an incorrect branch ( ` raise ValueError(f"unknown modification: {modification}`), and test is much easier to comprehend

A rough sketch:

```py
  _INIT = Initcode(deploy_code=Op.STOP)

  _NONCE_CODE = (
      Op.MSTORE(0, Op.PUSH32(bytes(_INIT)))
      + Op.CREATE(0, 32 - len(_INIT), len(_INIT))
      + Op.STOP
  )
  _STORE_CODE = Op.SSTORE(0x01, 0xCAFE) + Op.STOP


  @pytest.mark.with_all_create_opcodes
  @pytest.mark.pre_alloc_mutable()
  @pytest.mark.parametrize(
      "x_code, follow_up, x_bal, x_post, deploys_inner",
      [
          pytest.param(
              Op.STOP, False,
              BalAccountExpectation.empty(),
              Account(code=Op.STOP, nonce=1, balance=0, storage={}),
              False,
              id="collision_only",
          ),
          pytest.param(
              _NONCE_CODE, True,
              BalAccountExpectation(
                  nonce_changes=[BalNonceChange(block_access_index=2, post_nonce=2)],
              ),
              Account(code=_NONCE_CODE, nonce=2, balance=0, storage={}),
              True,
              id="then_nonce_change",
          ),
          pytest.param(
              _STORE_CODE, True,
              BalAccountExpectation(storage_changes=[BalStorageSlot(
                  slot=0x01,
                  slot_changes=[BalStorageChange(block_access_index=2, post_value=0xCAFE)],
              )]),
              Account(code=_STORE_CODE, nonce=1, balance=0, storage={0x01: 0xCAFE}),
              False,
              id="then_storage_change",
          ),
      ],
  )
  def test_bal_create_collision(
      pre, blockchain_test, fork, create_opcode,
      x_code, follow_up, x_bal, x_post, deploys_inner,
  ):
      alice = pre.fund_eoa()
      factory = pre.deploy_contract(code=_factory(create_opcode), storage={0: 0xDEAD})
      x = compute_create_address(address=factory, nonce=1, salt=0,
                                 initcode=bytes(_INIT), opcode=create_opcode)
      pre[x] = Account(code=x_code, nonce=1)

      gas = fork.transaction_gas_limit_cap()
      txs = [Transaction(sender=alice, to=factory, gas_limit=gas)]
      expectations = {alice: _alice_bal(), factory: _factory_bal(), x: x_bal}
      post = {alice: Account(nonce=1), factory: Account(nonce=2, storage={0: 0}), x: x_post}

      if follow_up:
          bob = pre.fund_eoa()
          txs.append(Transaction(sender=bob, to=x, gas_limit=gas))
          expectations[bob] = BalAccountExpectation(
              nonce_changes=[BalNonceChange(block_access_index=2, post_nonce=1)])
          post[bob] = Account(nonce=1)

      if deploys_inner:
          child = compute_create_address(address=x, nonce=1)
          expectations[child] = BalAccountExpectation(
              nonce_changes=[BalNonceChange(block_access_index=2, post_nonce=1)],
              code_changes=[BalCodeChange(block_access_index=2, new_code=bytes(Op.STOP))])
          post[child] = Account(nonce=1, code=bytes(Op.STOP), balance=0, storage={})

      blockchain_test(pre=pre, blocks=[Block(
          txs=txs,
          expected_block_access_list=BlockAccessListExpectation(account_expectations=expectations),
      )], post=post)
```

Comment (fselmo) [reply tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py:2839]: Thanks. I personally prefer the invariant approach and I feel like it's a bit more readable than having complex params. I'm going to skip it here if that's OK.

Comment (raxhvl) [reply tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py:2839]: Ofcourse!

File: tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py [hunk 3]
```diff
@@ -3022,6 +3085,117 @@ def test_bal_transient_storage_not_tracked(
...
+    via CREATE2, tx2 retries the same CREATE2 → collision. Covers the
+    `code_changes` axis of #2914 (forward order would need 7702 +
+    signable EOA at a deterministic CREATE address, infeasible).
+
+    CREATE2-only: CREATE auto-increments factory.nonce between txs, so
+    the second attempt targets a different address.
+    """
+    alice = pre.fund_eoa()
+
+    init_code = Initcode(deploy_code=Op.STOP)
```
Comment (raxhvl) [inline tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py]: Suggestion: Probe the interaction between init code and collusion

I *strongly* suggest having the init code increment storage on each invocation (`SSTORE(0, ADD(SLOAD(0), 1))`). Then assert:
- the storage slot [0] equals 1 (the second creation caused a collusion), and
- the colluding transaction at index 2 does not record a storage access. Init code MUST NOT be executed during a collusion; a buggy client might record a read by init + revert on collusion (demoted from a storage write).

Comment (fselmo) [reply tests/amsterdam/eip7928_block_level_access_lists/test_block_access_lists_opcodes.py]: Yes, good point, I will add this to the test.

**Review by raxhvl — APPROVED** [review]
Awesome. Good to merge.

**Comment by jangko** [comment]
~Is there any test covering case like: collision, then account get deleted in the same tx. nonce become x -> 0, code hash x -> EMPTY_CODEHASH?~

~Don't know what will happen with the storage in that situation.~

Don't mind it, it's not related to collision.

**Review by spencer-tb — APPROVED** [review]
LGTM! Thanks.

**Comment by spencer-tb** [comment]
First time rebasing manually... 
Screenshot 2026-05-28 at 13 23 29
